File: easy/101-200/187/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"strings"
)

type Flag struct {
	Short, Long string
}

func (f *Flag) IsActive(input string) bool {
	return f.Short == input || f.Long == input
}

func (f *Flag) String() string {
	return fmt.Sprintf("flag: %s", f.Long)
}

func NewFlag(short, long string) *Flag {
	return &Flag{
		Short: short,
		Long:  long,
	}
}

var zPrint = fmt.Println

func main() {
	_, _ = zPrint("hello world")
}

func parseFlags(input string, availableFlags []*Flag) (output string) {
	argv := strings.Split(input, " ")
	for _, arg := range argv {
		if strings.HasPrefix(arg, "--") {
			currentArg := strings.TrimPrefix(arg, "--")
			found := false
			for _, flag := range availableFlags {
				if flag.IsActive(currentArg) {
					output += flag.String() + "\n"
					found = true
					break
				}
			}
			if !found {
				output += fmt.Sprintf("flag: %s\n", currentArg)
			}
		} else if strings.HasPrefix(arg, "-") {
			for _, shortFlag := range strings.Split(strings.TrimPrefix(arg, "-"), "") {
				for _, flag := range availableFlags {
					if flag.IsActive(shortFlag) {
						output += flag.String() + "\n"
						break
					}
				}
			}
		} else {
			output += fmt.Sprintf("parameter: %s\n", arg)
		}
	}
	return
}

File: easy/101-200/187/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#[cfg(not(tarpaulin_include))]
fn main() {
    println!("rad");
}

fn parse_flags(available_flags: Vec<&str>, input: &str) -> Vec<String> {
    let mut flags = Vec::new();
    for chunk in input.split_whitespace() {
        if chunk.starts_with("--") {
            let flag = &chunk[2..];
            flags.push(format!("flag: {}", flag.to_string()));
        } else if chunk.starts_with("-") {
            for char in chunk[1..].chars() {
                let index = available_flags
                    .iter()
                    .position(|&x| x.to_string() == char.to_string())
                    .unwrap();
                flags.push(format!("flag: {}", available_flags[index + 1].to_string()));
            }
        } else {
            flags.push(format!("parameter: {}", chunk.to_string()));
        }
    }
    flags
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_flags() {
        let available_flags = vec![
            "a",
            "all",
            "f",
            "force",
            "n",
            "networking",
            "N",
            "numerical-list",
        ];
        let input = "-aN 12 --verbose 192.168.0.44";
        let expected = vec![
            "flag: all",
            "flag: numerical-list",
            "parameter: 12",
            "flag: verbose",
            "parameter: 192.168.0.44",
        ];
        assert_eq!(expected, parse_flags(available_flags, input));
    }
}
